File: 백준/Silver/1966. 프린터 큐/프린터 큐.py
import sys
import collections

test = int(sys.stdin.readline())

for i in range(test):
    N, M = map(int, sys.stdin.readline().split())
    priority = collections.deque(map(int, sys.stdin.readline().split()))
    chooseNum = M # 골라야하는 숫자의 번호를 저장
    count = 0 # 몇번째로 출력되는지 번호 저장
    while(len(priority) > 0):
        maxNum = max(priority) # 반복할때마다 우선순위가 제일 높은 값 저장
        if(priority[0] < maxNum): # 우선순위에서 밀린경우
            num = priority.popleft() # 맨왼쪽값빼기
            priority.append(num)
            if(chooseNum == 0): # 골라야하는 숫자의 번호가 0번이면 맨뒤로
                chooseNum = len(priority) - 1 
            else: # 아닌경우에는 자연스럽게 1빼기
                chooseNum -= 1
        elif(priority[0] == maxNum): # 우선순위인 경우
            priority.popleft() # 맨왼쪽 깂 제거 진행
            if(chooseNum == 0): # 이 값이 골라야하는 번호이면
                count += 1
                break # 반복문 탈출
            else:
                count += 1
                if(chooseNum == 0):
                    chooseNum = len(priority) - 1
                else:
                    chooseNum -= 1
    print(count)

    

File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
// created by Ipad
import java.util.*;
import java.io.*;

public class Main{

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static Deque<Integer> priority;



    public static void main(String[] args) throws IOException{

        int T = Integer.parseInt(br.readLine());

        for(int i = 0 ; i < T ; i++) {
            priority = new LinkedList<>();
            st = new StringTokenizer(br.readLine(), " ");
            int N = Integer.parseInt(st.nextToken());
            int M = Integer.parseInt(st.nextToken());
            st = new StringTokenizer(br.readLine(), " ");
            while(st.hasMoreTokens()) {
                priority.add(Integer.parseInt(st.nextToken()));
            }

           int count = 1;
            
            //peek 는 맨앞에 값 확인용, poll은 맨 앞에값 빼버리기
            while(true) {
                int maxPriority = Collections.max(priority);
                if(priority.peek() == maxPriority) {
                    if(M == 0) {
                        break;
                    }
                    else {
                        priority.poll();
                        M = ((M - 1) + priority.size()) % priority.size();
                        count += 1;
                    }
                }
                else {
                    int num = priority.peek();
                    priority.poll();
                    priority.addLast(num);
                    M = ((M - 1) + priority.size()) % priority.size();
                }   
            }
            bw.write(Integer.toString(count));
            bw.write("\n"); 
        
        }

        bw.flush();

        br.close();
        bw.close();
    }
}